core/simulations/helpers/createVaultWithCollateral.ts

```diff
@@ -186,4 +207,30 @@ const createVaultWithCollateral = async (collateralType: CollateralType, collate
     return vaultId;
 };
 
+const createVaultWithCollateral = async (collateralType: CollateralType, collateralOwned: BigNumber) => {
+    const collateralConfig = getCollateralConfigByType(collateralType);
+    const [balanceSlot, languageFormat] = await determineBalanceSlot(collateralType);
+    if (balanceSlot && languageFormat) {
+        await setCollateralInWallet(collateralConfig.symbol, collateralOwned);
+    } else {
+        // fallback to setting vat balance and withdrawing it
+        await setAndCheckCollateralInVat(collateralType, collateralOwned);
+        await checkAndWithdrawCollateralFromVat(collateralConfig, collateralOwned);
+    }
+    await ensureWalletBalance(collateralConfig, collateralOwned);
+
+    try {
+        return await createDefaultVaultWithCollateral(collateralType, collateralOwned);
+    } catch (e) {
+        if (e instanceof Error && e.message.includes('Error: Transaction reverted without a reason string')) {
```

Comment:
This error is very generic. I wonder if there is way to actually know that this collateral requires proxy eg `join.implementation` public value exist on the proxy-like contracts (since it's the address of the actual join contract it seems)? Then, you can just call `join.implementation` before trying to create default vault and not try/catch at all?

Reply:
By the way, why don't we never use `implementation` address anywhere? Isn't it the desired contract already with which we can open a vault?

Reply:
Using the library now.

> Isn't it the desired contract already with which we can open a vault?

no, here as far as i understand we're talking about the join contract. Join contract is responsible for holding funds and letting VAT know that the user has deposited the funds.

On top of that, neither of our ABIs contains the declaration of the `implementation` const (aka claims that no such const is present). The abi that we store for `join` of the crveth under the name `MCD_JOIN_CRVV1ETHSTETH_A` is in fact the abi of this contract's proxy. This allows us to know which join function we have access to in the implementation through the proxy, but does not inform us about the proxy's functions.

Reply:
> On top of that, neither of our ABIs contains the declaration of the implementation const

This is a limitation of our codebase, not a technical problem. You can either pass a different ABI in the `getContractInterfaceByName` for this contract (isolating away the detection of the contract type) to somewhere else; or build a simple human-readable ABI for this purpose yourself (eg `["function implementation() view returns (address)"]`). But anyway, you already use a library for this (to get the `proxyTarget` address)

> as far as i understand we're talking about the join contract

As opposite to what contract? I will repeat my question. When you find `proxyTarget` can't you just use it instead of creating a new a new proxy? Can you actually try it?

Reply:
> When you find proxyTarget can't you just use it instead of creating a new a new proxy?

the found proxy target is the JOIN contract.

In order to create the vault you require DS-PROXY contract. 

These 2 are used for different purpose.

see https://github.com/sidestream-tech/unified-auctions-ui/issues/501#issuecomment-1293491292 with overview of what-proxy-for-what.